### DISCUSSION TITLE
Error deploying to Rinkeby Test network with Metamask (End of Lesson 4)

### DISCUSSION THREAD

**Discussion by NewCoderrrrr** [smartcontractkit/full-blockchain-solidity-course-py#1188] [Q&A] [answered]
I would like to start off by thanking anyone in advance for help on my issue, also this is my first time posting to GitHub so the SimpleStorage.sol code seems to be messed up at the bottom. 

Anyways I'm at the very end of lesson 4. I saw another thread that someone made on this on January 17th, but I tried most of the recommendations in there and none of them worked. 

Alright let me get to my point: I got the integrated genache-cli working perfectly, but deploying to the blockchain Rinkeby Test network with my metamask, I am running into some issues. 

**This is the error I am getting:** 
Traceback (most recent call last):
  File "deploy.py", line 70, in <module>
    transaction = SimpleStorage.constructor().buildTransaction(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_utils/decorators.py", line 18, in _wrapper
    return self.method(obj, *args, **kwargs)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/contract.py", line 684, in buildTransaction
    return fill_transaction_defaults(self.web3, built_transaction)
  File "cytoolz/functoolz.pyx", line 250, in cytoolz.functoolz.curry.__call__
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/_utils/transactions.py", line 114, in fill_transaction_defaults
    default_val = default_getter(web3, transaction)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/_utils/transactions.py", line 60, in <lambda>
    'gas': lambda web3, tx: web3.eth.estimate_gas(tx),
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/eth.py", line 855, in estimate_gas
    return self._estimate_gas(transaction, block_identifier)
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/module.py", line 57, in caller
    result = w3.manager.request_blocking(method_str,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 198, in request_blocking
    return self.formatted_response(response,
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 171, in formatted_response
    raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'gas required exceeds allowance (19000)'}

Some side notes: 
- Im on MacOS
- Running Python version 3.8.9
- I installed Xcode through apple store (Recommended in other post)
- I installed cython using pip3 install Cython (Recommended in other post)

```
import json
import os

from solcx import compile_standard, install_solc

from web3 import Web3

from dotenv import load_dotenv

load_dotenv()

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

# print("Installing...")
install_solc("0.6.0")

# Solidity source code
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                }
            }
        },
    },
    solc_version="0.6.0",
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

# get bytecode
bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

# get abi
abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

# For connecting to ganache
# w3 = Web3(Web3.HTTPProvider("HTTP://127.0.0.1:8545"))

# For connecting with Rinkeby through infura
w3 = Web3(
    Web3.HTTPProvider("https://rinkeby.infura.io/v3/c1f022a4631242708ae275d213405e08")
)

chain_id = 4
my_address = "0x90F8bf6A479f320ead074411a4B0e7944Ea8c9C1"
# never hard code private key, type in terminal export PRIVATE_KEY= then the key
private_key = os.getenv("PRIVATE_KEY")

# Create the contract in Python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)

# Get the latest transaction
nonce = w3.eth.getTransactionCount(my_address)

# Steps:
# 1. Build a transaction
# 2. Sign a transaction
# 3. Send a transaction

# 1. Build transaction
transaction = SimpleStorage.constructor().buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce,
    }
)
# 2.Sign Transaction
signed_txn = w3.eth.account.sign_transaction(transaction, private_key=private_key)

# 3. Send this signed transaction
print("Deploying Contract")
tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
print("Deployed!")
# Working with the contract you always need a contract address and a contract abi

simple_storage = w3.eth.contract(address=tx_receipt.contractAddress, abi=abi)

# when making transactions on the block chain there are two ways to interact with them
# call -> simulate getting a call and getting a return value
# transaction -> actuallyy make a state change

# initial value of favorite number
print(simple_storage.functions.retrieve().call())
print("Updating Contract")
store_transaction = simple_storage.functions.store(15).buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce + 1,
    }
)

# sign it
signed_store_txn = w3.eth.account.sign_transaction(
    store_transaction, private_key=private_key
)
# send transaction
send_store_tx = w3.eth.send_raw_transaction(signed_store_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(send_store_tx)
print("Updated!")
print(simple_storage.functions.retrieve().call())

```
``
pragma solidity ^0.6.0;

contract SimpleStorage {
    // this will get initialized to 0!
    uint256 favoriteNumber;
    bool favoriteBool;

    struct People {
        uint256 favoriteNumber;
        string name;
    }

    People[] public people;
    mapping(string => uint256) public nameToFavoriteNumber;

    function store(uint256 _favoriteNumber) public returns (uint256) {
        favoriteNumber = _favoriteNumber;
        return _favoriteNumber;
    }

    function retrieve() public view returns (uint256) {
        return favoriteNumber;
    }

    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }
}
``

**Comment by charlesAuthentik** [comment]
Do you have enough ETH on your Rinkeby account?

**Comment by cromewar** [comment]
Hello @NewCoderrrrr you must change the gas limit parameter, if you are using ganache UI you can change that on the settings, and you you are using ganache-cli you can change the gas limit on the transaction adding it manually, also check this:
https://eth-brownie.readthedocs.io/en/stable/api-network.html#main.gas_limit

**Reply by NewCoderrrrr** [reply]
Thanks @cromewar I will try that right now and get back to you on whether or not it worked. Appreciate it!

**Comment by NewCoderrrrr** [comment]
@cromewar I hate to admit this but my above problem ended up being that I was not using the right metamask address, but once I fixed that issue I have ran into a few new ones. It is weird because the more and more I run this program I run into new error's, i've provided the 3 I have received so far.

Side Notes:

-  I never ended up changing the gas price as you previously requested because I didn't know how to. I'd assume it's not related to these new errors I don't think because what was causing that error was me not having the right metamask address linked.
- I read some other discussion thread that was some what related to issues i'm encountering and they said they fixed it by importing the SimpleStorage.sol provided in the courses github. I changed mine to the one from the GitHub and then I got the 3rd error.
- Also I keep seeing you refer to running python on a venv. I don't know how to do that or whether or not I am already set up that way. 

1st Error:
Deploying Contract
Traceback (most recent call last):
File "deploy.py", line 83, in
tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/eth.py", line 818, in send_raw_transaction
return self._send_raw_transaction(transaction)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/module.py", line 57, in caller
result = w3.manager.request_blocking(method_str,
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 198, in request_blocking
return self.formatted_response(response,
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 171, in formatted_response
raise ValueError(response["error"])
ValueError: {'code': -32000, 'message': 'nonce too low'}

2nd Error:
Deploying Contract
Deployed!
Traceback (most recent call last):
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/contract.py", line 1513, in call_contract_function
output_data = web3.codec.decode_abi(output_types, return_data)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/codec.py", line 181, in decode_abi
return decoder(stream)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/decoding.py", line 127, in call
return self.decode(stream)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_utils/functional.py", line 45, in inner
return callback(fn(*args, **kwargs))
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/decoding.py", line 173, in decode
yield decoder(stream)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/decoding.py", line 127, in call
return self.decode(stream)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/decoding.py", line 198, in decode
raw_data = self.read_data_from_stream(stream)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/eth_abi/decoding.py", line 305, in read_data_from_stream
raise InsufficientDataBytes(
eth_abi.exceptions.InsufficientDataBytes: Tried to read 32 bytes. Only got 0 bytes

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "deploy.py", line 95, in
print(simple_storage.functions.retrieve().call())
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/contract.py", line 957, in call
return call_contract_function(
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/contract.py", line 1530, in call_contract_function
raise BadFunctionCallOutput(msg) from e
web3.exceptions.BadFunctionCallOutput: Could not transact with/call contract function, is contract deployed correctly and chain synced?

**__3rd Error:__**
File "deploy.py", line 62, in
nonce = w3.eth.getTransactionCount(my_address)
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/module.py", line 57, in caller
result = w3.manager.request_blocking(method_str,
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 198, in request_blocking
return self.formatted_response(response,
File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/web3/manager.py", line 171, in formatted_response
raise ValueError(response["error"])
ValueError: {'code': -32603, 'message': 'request failed or timed out'}

**Reply by cromewar** [reply]
let me do more research on this.

**Reply by NewCoderrrrr** [reply]
@cromewar 
So I was able to fix it but I would still like some clarification, under this text I have my own code. I finally decided to just "clone the repo" straight from the GitHub [My first time ever doing this]. I then changed the cloned code to include my own information. This worked!!!

I am trying to figure out what made it work however so that I know moving forward. I noticed some "requirements.txt" file that had 

web3==5.23.1
py-solc-x 

in it, which I changed to 
web3==5.28.0
py-solc-x 

because 5.28.0 is the most recent version. I then noticed something in the read.me text that said to type "pip3 install -r requirements.txt". I did this and it seemed to uninstall and reinstall some things. Is this what made my code work because I had the "updated" web3??? If this is true, do I have to do this every time I start a new project from here on out or is typing "pip3 install -r requirements.txt" a one time thing?

There are two other things that are different in my code that I copied while watching the video, compared with the code that came when I took the repo: 
1.) At the bottom of my deploy.py code starting at the line "store_transaction = simple_storage.functions.store(15).buildTransaction(" every time I said "store" the repo had "greeting" such as "greeting_transaction". In my mind I would assume this doesn't matter.
2.) In my code from the video I have: 
```
# get abi
abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]
```
but in the repo it has:
```
abi = json.loads(
    compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["metadata"]
)["output"]["abi"]
```
Why is it that this has json.loads but bytecode doesn't in the repo? also "["metadata"]
)["output"]" 

Again this was my original code, not the code that it ended up working with: 
⬇️⬇️⬇️

SimpleStorage.sol
```
// SPDX-License-Identifier: MIT

pragma solidity >=0.6.0 <0.9.0;

contract SimpleStorage {
    uint256 favoriteNumber;

    // This is a comment!
    struct People {
        uint256 favoriteNumber;
        string name;
    }

    People[] public people;
    mapping(string => uint256) public nameToFavoriteNumber;

    function store(uint256 _favoriteNumber) public {
        favoriteNumber = _favoriteNumber;
    }

    function retrieve() public view returns (uint256) {
        return favoriteNumber;
    }

    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }
}
```
deploy.py
```
import json

from web3 import Web3

from solcx import compile_standard, install_solc
import os
from dotenv import load_dotenv

load_dotenv()

with open("./SimpleStorage.sol", "r") as file:
    simple_storage_file = file.read()

# We add these two lines that we forgot from the video!
# print("Installing...")
install_solc("0.6.0")

# Solidity source code
compiled_sol = compile_standard(
    {
        "language": "Solidity",
        "sources": {"SimpleStorage.sol": {"content": simple_storage_file}},
        "settings": {
            "outputSelection": {
                "*": {
                    "*": ["abi", "metadata", "evm.bytecode", "evm.bytecode.sourceMap"]
                }
            }
        },
    },
    solc_version="0.6.0",
)

with open("compiled_code.json", "w") as file:
    json.dump(compiled_sol, file)

# get bytecode
bytecode = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["evm"][
    "bytecode"
]["object"]

# get abi
abi = compiled_sol["contracts"]["SimpleStorage.sol"]["SimpleStorage"]["abi"]

# For connecting to ganache
# w3 = Web3(Web3.HTTPProvider("HTTP://127.0.0.1:8545"))

# For connecting with Rinkeby through infura
w3 = Web3(
    Web3.HTTPProvider("https://rinkeby.infura.io/v3/c1f022a4631242708ae275d213405e08")
)

chain_id = 4
my_address = "0xA0ca53205Aa888dFbF25642F501314Df1760a20e"

# never hard code private key unless in .env
private_key = os.getenv("PRIVATE_KEY")

# Create the contract in Python
SimpleStorage = w3.eth.contract(abi=abi, bytecode=bytecode)

# Get the latest transaction
nonce = w3.eth.getTransactionCount(my_address) + 20

# Steps:
# 1. Build a transaction
# 2. Sign a transaction
# 3. Send a transaction

# 1. Build transaction
transaction = SimpleStorage.constructor().buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce,
    }
)
# 2.Sign Transaction
signed_txn = w3.eth.account.sign_transaction(transaction, private_key=private_key)

# 3. Send this signed transaction
print("Deploying Contract")
tx_hash = w3.eth.send_raw_transaction(signed_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(tx_hash)
print("Deployed!")
# Working with the contract you always need a contract address and a contract abi

simple_storage = w3.eth.contract(address=tx_receipt.contractAddress, abi=abi)

# when making transactions on the block chain there are two ways to interact with them
# call -> simulate getting a call and getting a return value
# transaction -> actuallyy make a state change

# initial value of favorite number
print(simple_storage.functions.retrieve().call())
print("Updating Contract")
store_transaction = simple_storage.functions.store(15).buildTransaction(
    {
        "chainId": chain_id,
        "gasPrice": w3.eth.gas_price,
        "from": my_address,
        "nonce": nonce + 1,
    }
)

# sign it
signed_store_txn = w3.eth.account.sign_transaction(
    store_transaction, private_key=private_key
)
# send transaction
send_store_tx = w3.eth.send_raw_transaction(signed_store_txn.rawTransaction)
tx_receipt = w3.eth.wait_for_transaction_receipt(send_store_tx)
print("Updated!")
print(simple_storage.functions.retrieve().call())

```
.env 
```
export PRIVATE_KEY=My own (In actual code its right)
# have to run in terminal: source .env 
# to update private key whenever it is changed
# apparently only on maac
```
.gitignore
`.env`

**Reply by n4n0b1t3** [reply]
@cromewar A good example of why a series of short articles about how to understand errors and fix them would be a good idea. This would be the use case, where we talk about finding differences between working code (from the repo) and not working code. Here for example a notepad++ plugin which could be applied by windows users: [How to Make Notepad++ Compare Two Files With a Plugin](https://www.makeuseof.com/tag/notepad-compare-two-files-plugin/#:~:text=One%20such%20powerful%20plugin%20is,difference%20between%20the%20two%20files.)

**Comment by cromewar** [comment, marked as answer]
Hello @NewCoderrrrr 

> I'll try to answer you questions:
> because 5.28.0 is the most recent version. I then noticed something in the read.me text that said to type "pip3 install -r requirements.txt". I did this and it seemed to uninstall and reinstall some things. Is this what made my code work because I had the "updated" web3??? If this is true, do I have to do this every time I start a new project from here on out or is typing "pip3 install -r requirements.txt" a one time thing?

I depends, if the project repo you are working on on have the requirements.txt is always a good practice to use `pip3 install -r requirements` but please do this on a virtualenv not in your main python installation. the reason it uninstalled and reinstalled things is due to the versions, each version of a package could have different dependencies with also different versions. Also, take in mind that is not 100% necessary to use always the latest version of a package, using the one on requirements.txt should be enough.

Note: Do not forget to use python virtual env.

> Why is it that this has json.loads but bytecode doesn't in the repo? also "["metadata"]
> )["output"]"

`json.loads` is a decoder for json files, basically a standard to organize a json file, you can see more info [here](https://docs.python.org/3/library/json.html), so, when you are interacting with a contract you need two things, and ABI and the contract address. The bytecode is not needed to be encoded as you are not going to use it, but the ABI is as you need it to interact with the contract. 

Please feel free to make any other questions ;)

**Reply by NewCoderrrrr** [reply]
Ok the explanation on the json.loads makes sense. It was just confusing because it wasn't in the video. 

Speaking on running Python and Visual code in a virtual environment, is this common practice? How does one run it as aa virtual environment? Do I need a virtual machine and then run linux on it or something like that? Or is there a much easier way? 

Thanks for the help!

**Reply by cromewar** [reply]
Not at all, virtualenv is just a package for python wich allows you to create isolated instances of python, follow [this](https://gist.github.com/Geoyi/d9fab4f609e9f75941946be45000632b) guide. Once you have the virtualenv created you should select it on vscode (bottom left corner) and you are ready to go.

**Reply by NewCoderrrrr** [reply]
Thank you so much @cromewar
